This patch touches both the `requirements.txt` which is used for installation
during the package build and, due to the complexity of the requirements graph,
the `requirements.in` file used by upstream to generate `requirements.txt`.
Often updating `requirements.in` and regenerating `requirements.txt` will be
required to resolve dependency issues.

In order to do that, add the following lines to the top of the `runs` pipeline
stanza, which will install `pip-compile` and run it once to populate the local
pip cache, then fail the build to give you a debug shell:

    apk add py${{vars.py-version}}-pip-tools-bin
    pip-compile --output-file - backend/src/apiserver/visualization/requirements.in
    false

Make the required changes to `requirements.in` and then run:

    pip-compile --output-file tmp.txt backend/src/apiserver/visualization/requirements.in
    mv tmp.txt backend/src/apiserver/visualization/requirements.txt

(Generating to `tmp.txt` and moving into place is required to avoid
`pip-compile` using the existing `requirements.txt` content to inform its
decisions.)

Once generated, run `git diff` and use its output to replace the below content.

---
diff --git a/backend/src/apiserver/visualization/requirements.in b/backend/src/apiserver/visualization/requirements.in
index e24773c..46be981 100644
--- a/backend/src/apiserver/visualization/requirements.in
+++ b/backend/src/apiserver/visualization/requirements.in
@@ -1,18 +1,20 @@
-bokeh==1.2.0
+bokeh==2.3.2  # required for support for Jinja2 version
+cryptography>=44.0.1,<45
 gcsfs==0.2.3
 google-api-python-client==1.7.*
 itables==0.1.0
-ipykernel==5.1.1
-ipython==7.12.0
+ipykernel==6.29.5
+ipython==7.34.0
 Jinja2>=3.1.2,<4
 jupyter-client==8.6.3
-markupsafe==2.0.1
 nbconvert==7.16.5
 nbformat==5.10.4
-scikit-learn==0.24.2
-tensorflow==2.10.1
-tensorflow-metadata==1.9.*
-tensorflow-model-analysis==0.40.*
-tensorflow-data-validation==1.9.*
-tensorflow-serving-api==2.10.1
+notebook==7.3.2  # CVE-2024-43805 GHSA-9q39-rmj3-p4r2
+scikit-learn==1.5.0
+tensorflow==2.16.2
+tensorflow-metadata==1.16.1
+tensorflow-model-analysis==0.47.*
+tensorflow-data-validation==1.16.1
+tensorflow-serving-api==2.14.1
 tornado==6.*
+werkzeug==3.0.6  # CVE-2024-49766 GHSA-f9vj-2wh5-fj8j, CVE-2024-49767 GHSA-q34m-jh98-gwm2
diff --git a/backend/src/apiserver/visualization/requirements.txt b/backend/src/apiserver/visualization/requirements.txt
index fa15911..641e5ff 100644
--- a/backend/src/apiserver/visualization/requirements.txt
+++ b/backend/src/apiserver/visualization/requirements.txt
@@ -1,24 +1,27 @@
 #
-# This file is autogenerated by pip-compile with Python 3.8
+# This file is autogenerated by pip-compile with Python 3.11
 # by the following command:
 #
-#    pip-compile --output-file=- --resolver=backtracking -
+#    pip-compile --output-file=tmp.txt backend/src/apiserver/visualization/requirements.in
 #
 absl-py==1.4.0
     # via
+    #   keras
+    #   rouge-score
     #   tensorboard
     #   tensorflow
     #   tensorflow-data-validation
     #   tensorflow-metadata
     #   tensorflow-model-analysis
     #   tfx-bsl
-anyio==4.5.2
+annotated-types==0.7.0
+    # via pydantic
+anyio==4.8.0
     # via
     #   httpx
     #   jupyter-server
-apache-beam[gcp]==2.46.0
+apache-beam[gcp]==2.63.0
     # via
-    #   apache-beam
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
     #   tfx-bsl
@@ -32,54 +35,74 @@ astunparse==1.6.3
     # via tensorflow
 async-lru==2.0.4
     # via jupyterlab
-attrs==24.3.0
+attrs==25.1.0
     # via
     #   jsonschema
     #   referencing
-babel==2.16.0
+babel==2.17.0
     # via jupyterlab-server
 backcall==0.2.0
     # via ipython
-beautifulsoup4==4.12.3
+backports-tarfile==1.2.0
+    # via jaraco-context
+beautifulsoup4==4.13.3
     # via nbconvert
-bleach[css]==6.1.0
-    # via
-    #   bleach
-    #   nbconvert
-bokeh==1.2.0
-    # via -r -
-cachetools==4.2.4
+bleach[css]==6.2.0
+    # via nbconvert
+bokeh==2.3.2
+    # via -r backend/src/apiserver/visualization/requirements.in
+cachetools==5.5.2
     # via
     #   apache-beam
     #   google-auth
-certifi==2024.12.14
+certifi==2025.1.31
     # via
     #   httpcore
     #   httpx
     #   requests
 cffi==1.17.1
-    # via argon2-cffi-bindings
+    # via
+    #   argon2-cffi-bindings
+    #   cryptography
 charset-normalizer==3.4.1
     # via requests
+click==8.1.8
+    # via nltk
 cloudpickle==2.2.1
     # via apache-beam
+colorama==0.4.6
+    # via sacrebleu
 comm==0.2.2
-    # via ipywidgets
+    # via
+    #   ipykernel
+    #   ipywidgets
 crcmod==1.7
     # via apache-beam
+cryptography==44.0.1
+    # via
+    #   -r backend/src/apiserver/visualization/requirements.in
+    #   secretstorage
+debugpy==1.8.12
+    # via ipykernel
 decorator==5.1.1
     # via
     #   gcsfs
     #   ipython
 defusedxml==0.7.1
     # via nbconvert
+deprecated==1.2.18
+    # via
+    #   opentelemetry-api
+    #   opentelemetry-semantic-conventions
 dill==0.3.1.1
     # via apache-beam
+dnspython==2.7.0
+    # via pymongo
 docopt==0.6.2
     # via hdfs
-exceptiongroup==1.2.2
-    # via anyio
-fastavro==1.9.7
+docstring-parser==0.16
+    # via google-cloud-aiplatform
+fastavro==1.10.0
     # via apache-beam
 fasteners==0.19
     # via
@@ -87,17 +110,18 @@ fasteners==0.19
     #   google-apitools
 fastjsonschema==2.21.1
     # via nbformat
-flatbuffers==24.12.23
+flatbuffers==25.2.10
     # via tensorflow
 fqdn==1.5.1
     # via jsonschema
-gast==0.4.0
+gast==0.6.0
     # via tensorflow
 gcsfs==0.2.3
-    # via -r -
-google-api-core[grpc]==2.24.0
+    # via -r backend/src/apiserver/visualization/requirements.in
+google-api-core[grpc]==2.24.1
     # via
-    #   google-api-core
+    #   apache-beam
+    #   google-cloud-aiplatform
     #   google-cloud-bigquery
     #   google-cloud-bigquery-storage
     #   google-cloud-bigtable
@@ -108,16 +132,18 @@ google-api-core[grpc]==2.24.0
     #   google-cloud-pubsub
     #   google-cloud-pubsublite
     #   google-cloud-recommendations-ai
+    #   google-cloud-resource-manager
     #   google-cloud-spanner
+    #   google-cloud-storage
     #   google-cloud-videointelligence
     #   google-cloud-vision
 google-api-python-client==1.7.12
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   tfx-bsl
 google-apitools==0.5.31
     # via apache-beam
-google-auth==2.37.0
+google-auth==2.38.0
     # via
     #   apache-beam
     #   gcsfs
@@ -125,73 +151,96 @@ google-auth==2.37.0
     #   google-api-python-client
     #   google-auth-httplib2
     #   google-auth-oauthlib
+    #   google-cloud-aiplatform
     #   google-cloud-bigquery
+    #   google-cloud-bigquery-storage
+    #   google-cloud-bigtable
     #   google-cloud-core
+    #   google-cloud-datastore
     #   google-cloud-dlp
+    #   google-cloud-language
     #   google-cloud-pubsub
+    #   google-cloud-recommendations-ai
+    #   google-cloud-resource-manager
+    #   google-cloud-storage
+    #   google-cloud-videointelligence
     #   google-cloud-vision
-    #   tensorboard
-google-auth-httplib2==0.1.1
+    #   keyrings-google-artifactregistry-auth
+google-auth-httplib2==0.2.0
     # via
     #   apache-beam
     #   google-api-python-client
-google-auth-oauthlib==0.4.6
-    # via
-    #   gcsfs
-    #   tensorboard
-google-cloud-bigquery==3.27.0
+google-auth-oauthlib==1.2.1
+    # via gcsfs
+google-cloud-aiplatform==1.81.0
     # via apache-beam
-google-cloud-bigquery-storage==2.16.2
+google-cloud-bigquery==3.29.0
+    # via
+    #   apache-beam
+    #   google-cloud-aiplatform
+google-cloud-bigquery-storage==2.28.0
     # via apache-beam
-google-cloud-bigtable==1.7.3
+google-cloud-bigtable==2.28.1
     # via apache-beam
-google-cloud-core==2.4.1
+google-cloud-core==2.4.2
     # via
     #   apache-beam
     #   google-cloud-bigquery
     #   google-cloud-bigtable
     #   google-cloud-datastore
     #   google-cloud-spanner
-google-cloud-datastore==1.15.5
+    #   google-cloud-storage
+google-cloud-datastore==2.20.2
     # via apache-beam
-google-cloud-dlp==3.18.0
+google-cloud-dlp==3.27.0
     # via apache-beam
-google-cloud-language==1.3.2
+google-cloud-language==2.17.0
     # via apache-beam
-google-cloud-pubsub==2.21.4
+google-cloud-pubsub==2.28.0
     # via
     #   apache-beam
     #   google-cloud-pubsublite
 google-cloud-pubsublite==1.11.1
     # via apache-beam
-google-cloud-recommendations-ai==0.7.1
+google-cloud-recommendations-ai==0.10.16
     # via apache-beam
-google-cloud-spanner==3.46.0
+google-cloud-resource-manager==1.14.1
+    # via google-cloud-aiplatform
+google-cloud-spanner==3.52.0
     # via apache-beam
-google-cloud-videointelligence==1.16.3
+google-cloud-storage==2.19.0
+    # via
+    #   apache-beam
+    #   google-cloud-aiplatform
+google-cloud-videointelligence==2.16.0
     # via apache-beam
-google-cloud-vision==3.7.2
+google-cloud-vision==3.10.0
     # via apache-beam
-google-crc32c==1.5.0
-    # via google-resumable-media
+google-crc32c==1.6.0
+    # via
+    #   google-cloud-storage
+    #   google-resumable-media
 google-pasta==0.2.0
     # via tensorflow
 google-resumable-media==2.7.2
-    # via google-cloud-bigquery
-googleapis-common-protos[grpc]==1.63.1
+    # via
+    #   google-cloud-bigquery
+    #   google-cloud-storage
+googleapis-common-protos[grpc]==1.68.0
     # via
     #   google-api-core
     #   grpc-google-iam-v1
     #   grpcio-status
     #   tensorflow-metadata
-grpc-google-iam-v1==0.12.7
+grpc-google-iam-v1==0.14.0
     # via
     #   google-cloud-bigtable
     #   google-cloud-pubsub
+    #   google-cloud-resource-manager
     #   google-cloud-spanner
 grpc-interceptor==0.15.4
     # via google-cloud-spanner
-grpcio==1.69.0
+grpcio==1.65.5
     # via
     #   apache-beam
     #   google-api-core
@@ -204,20 +253,22 @@ grpcio==1.69.0
     #   tensorboard
     #   tensorflow
     #   tensorflow-serving-api
-grpcio-status==1.48.2
+grpcio-status==1.63.0rc1
     # via
     #   google-api-core
     #   google-cloud-pubsub
     #   google-cloud-pubsublite
 h11==0.14.0
     # via httpcore
-h5py==3.11.0
-    # via tensorflow
+h5py==3.13.0
+    # via
+    #   keras
+    #   tensorflow
 hdfs==2.7.3
     # via apache-beam
 httpcore==1.0.7
     # via httpx
-httplib2==0.21.0
+httplib2==0.22.0
     # via
     #   apache-beam
     #   google-api-python-client
@@ -234,24 +285,15 @@ idna==3.10
     #   requests
 importlib-metadata==8.5.0
     # via
-    #   jupyter-client
-    #   jupyter-lsp
-    #   jupyterlab
-    #   jupyterlab-server
-    #   markdown
-    #   nbconvert
-importlib-resources==6.4.5
-    # via
-    #   jsonschema
-    #   jsonschema-specifications
-    #   jupyterlab
-ipykernel==5.1.1
+    #   keyring
+    #   opentelemetry-api
+ipykernel==6.29.5
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   jupyterlab
-ipython==7.12.0
+ipython==7.34.0
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   ipykernel
     #   ipywidgets
     #   tensorflow-model-analysis
@@ -262,51 +304,66 @@ ipywidgets==7.8.5
 isoduration==20.11.0
     # via jsonschema
 itables==0.1.0
-    # via -r -
+    # via -r backend/src/apiserver/visualization/requirements.in
+jaraco-classes==3.4.0
+    # via keyring
+jaraco-context==6.0.1
+    # via keyring
+jaraco-functools==4.1.0
+    # via keyring
 jedi==0.19.2
     # via ipython
-jinja2==3.1.5
+jeepney==0.8.0
+    # via
+    #   keyring
+    #   secretstorage
+jinja2==3.1.6
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   bokeh
     #   jupyter-server
     #   jupyterlab
     #   jupyterlab-server
     #   nbconvert
-joblib==0.14.1
+joblib==1.4.2
     # via
+    #   nltk
     #   scikit-learn
     #   tensorflow-data-validation
 json5==0.10.0
     # via jupyterlab-server
+jsonpickle==3.4.2
+    # via apache-beam
 jsonpointer==3.0.0
     # via jsonschema
 jsonschema[format-nongpl]==4.23.0
     # via
+    #   apache-beam
     #   jupyter-events
     #   jupyterlab-server
     #   nbformat
-jsonschema-specifications==2023.12.1
+jsonschema-specifications==2024.10.1
     # via jsonschema
 jupyter-client==8.6.3
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   ipykernel
     #   jupyter-server
     #   nbclient
 jupyter-core==5.7.2
     # via
+    #   ipykernel
     #   jupyter-client
     #   jupyter-server
     #   jupyterlab
     #   nbclient
     #   nbconvert
     #   nbformat
-jupyter-events==0.10.0
+jupyter-events==0.12.0
     # via jupyter-server
 jupyter-lsp==2.2.5
     # via jupyterlab
-jupyter-server==2.14.2
+jupyter-server==2.15.0
     # via
     #   jupyter-lsp
     #   jupyterlab
@@ -315,7 +372,7 @@ jupyter-server==2.14.2
     #   notebook-shim
 jupyter-server-terminals==0.5.3
     # via jupyter-server
-jupyterlab==4.1.5
+jupyterlab==4.3.5
     # via notebook
 jupyterlab-pygments==0.3.0
     # via nbconvert
@@ -325,49 +382,81 @@ jupyterlab-server==2.27.3
     #   notebook
 jupyterlab-widgets==1.1.11
     # via ipywidgets
-keras==2.10.0
-    # via tensorflow
-keras-preprocessing==1.1.2
+keras==3.9.0
     # via tensorflow
+keyring==25.6.0
+    # via keyrings-google-artifactregistry-auth
+keyrings-google-artifactregistry-auth==1.1.2
+    # via apache-beam
 libclang==18.1.1
     # via tensorflow
+lxml==5.3.1
+    # via sacrebleu
 markdown==3.7
     # via tensorboard
-markupsafe==2.0.1
+markdown-it-py==3.0.0
+    # via rich
+markupsafe==3.0.2
     # via
-    #   -r -
     #   jinja2
     #   nbconvert
-mistune==3.1.0
+    #   werkzeug
+matplotlib-inline==0.1.7
+    # via
+    #   ipykernel
+    #   ipython
+mdurl==0.1.2
+    # via markdown-it-py
+mistune==3.1.2
     # via nbconvert
-nbclient==0.10.1
+ml-dtypes==0.3.2
+    # via
+    #   keras
+    #   tensorflow
+more-itertools==10.6.0
+    # via
+    #   jaraco-classes
+    #   jaraco-functools
+namex==0.0.8
+    # via keras
+nbclient==0.10.2
     # via nbconvert
 nbconvert==7.16.5
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   jupyter-server
 nbformat==5.10.4
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   jupyter-server
     #   nbclient
     #   nbconvert
-notebook==7.1.3
-    # via widgetsnbextension
+nest-asyncio==1.6.0
+    # via ipykernel
+nltk==3.9.1
+    # via rouge-score
+notebook==7.3.2
+    # via
+    #   -r backend/src/apiserver/visualization/requirements.in
+    #   widgetsnbextension
 notebook-shim==0.2.4
     # via
     #   jupyterlab
     #   notebook
-numpy==1.24.4
+numpy==1.26.4
     # via
     #   apache-beam
     #   bokeh
     #   h5py
-    #   keras-preprocessing
+    #   keras
+    #   ml-dtypes
     #   pandas
     #   pyarrow
+    #   rouge-score
+    #   sacrebleu
     #   scikit-learn
     #   scipy
+    #   shapely
     #   tensorboard
     #   tensorflow
     #   tensorflow-data-validation
@@ -377,11 +466,22 @@ oauth2client==4.1.3
     # via google-apitools
 oauthlib==3.2.2
     # via requests-oauthlib
-objsize==0.6.1
+objsize==0.7.1
     # via apache-beam
+opentelemetry-api==1.30.0
+    # via
+    #   google-cloud-pubsub
+    #   opentelemetry-sdk
+    #   opentelemetry-semantic-conventions
+opentelemetry-sdk==1.30.0
+    # via google-cloud-pubsub
+opentelemetry-semantic-conventions==0.51b0
+    # via opentelemetry-sdk
 opt-einsum==3.4.0
     # via tensorflow
-orjson==3.10.14
+optree==0.14.0
+    # via keras
+orjson==3.10.15
     # via apache-beam
 overrides==7.7.0
     # via
@@ -389,11 +489,16 @@ overrides==7.7.0
     #   jupyter-server
 packaging==24.2
     # via
+    #   apache-beam
     #   bokeh
+    #   google-cloud-aiplatform
     #   google-cloud-bigquery
+    #   ipykernel
+    #   jupyter-events
     #   jupyter-server
     #   jupyterlab
     #   jupyterlab-server
+    #   keras
     #   nbconvert
     #   tensorflow
 pandas==1.5.3
@@ -410,30 +515,41 @@ pexpect==4.9.0
     # via ipython
 pickleshare==0.7.5
     # via ipython
-pillow==10.4.0
-    # via bokeh
-pkgutil-resolve-name==1.3.10
-    # via jsonschema
+pillow==11.1.0
+    # via
+    #   bokeh
+    #   tensorflow-model-analysis
 platformdirs==4.3.6
     # via jupyter-core
+pluggy==1.5.0
+    # via keyrings-google-artifactregistry-auth
+portalocker==3.1.1
+    # via sacrebleu
 prometheus-client==0.21.1
     # via jupyter-server
-prompt-toolkit==3.0.48
+prompt-toolkit==3.0.50
     # via ipython
-proto-plus==1.25.0
+proto-plus==1.26.0
     # via
     #   apache-beam
     #   google-api-core
+    #   google-cloud-aiplatform
     #   google-cloud-bigquery-storage
+    #   google-cloud-bigtable
+    #   google-cloud-datastore
     #   google-cloud-dlp
+    #   google-cloud-language
     #   google-cloud-pubsub
     #   google-cloud-recommendations-ai
+    #   google-cloud-resource-manager
     #   google-cloud-spanner
+    #   google-cloud-videointelligence
     #   google-cloud-vision
-protobuf==3.19.6
+protobuf==4.25.6
     # via
     #   apache-beam
     #   google-api-core
+    #   google-cloud-aiplatform
     #   google-cloud-bigquery-storage
     #   google-cloud-bigtable
     #   google-cloud-datastore
@@ -441,6 +557,7 @@ protobuf==3.19.6
     #   google-cloud-language
     #   google-cloud-pubsub
     #   google-cloud-recommendations-ai
+    #   google-cloud-resource-manager
     #   google-cloud-spanner
     #   google-cloud-videointelligence
     #   google-cloud-vision
@@ -455,16 +572,20 @@ protobuf==3.19.6
     #   tensorflow-model-analysis
     #   tensorflow-serving-api
     #   tfx-bsl
+psutil==7.0.0
+    # via ipykernel
 ptyprocess==0.7.0
     # via
     #   pexpect
     #   terminado
-pyarrow==5.0.0
+pyarrow==10.0.1
     # via
     #   apache-beam
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
     #   tfx-bsl
+pyarrow-hotfix==0.6
+    # via apache-beam
 pyasn1==0.6.1
     # via
     #   oauth2client
@@ -476,6 +597,10 @@ pyasn1-modules==0.4.1
     #   oauth2client
 pycparser==2.22
     # via cffi
+pydantic==2.10.6
+    # via google-cloud-aiplatform
+pydantic-core==2.27.2
+    # via pydantic
 pydot==1.4.2
     # via apache-beam
 pyfarmhash==0.3.2
@@ -484,9 +609,10 @@ pygments==2.19.1
     # via
     #   ipython
     #   nbconvert
-pymongo==3.13.0
+    #   rich
+pymongo==4.11.1
     # via apache-beam
-pyparsing==3.1.4
+pyparsing==3.2.1
     # via
     #   httplib2
     #   pydot
@@ -500,36 +626,44 @@ python-dateutil==2.9.0.post0
     #   pandas
 python-json-logger==3.2.1
     # via jupyter-events
-pytz==2024.2
+pytz==2025.1
     # via
     #   apache-beam
-    #   babel
     #   pandas
 pyyaml==6.0.2
     # via
+    #   apache-beam
     #   bokeh
     #   jupyter-events
-pyzmq==26.2.0
+pyzmq==26.2.1
     # via
+    #   ipykernel
     #   jupyter-client
     #   jupyter-server
-referencing==0.35.1
+redis==5.2.1
+    # via apache-beam
+referencing==0.36.2
     # via
     #   jsonschema
     #   jsonschema-specifications
     #   jupyter-events
 regex==2024.11.6
-    # via apache-beam
+    # via
+    #   apache-beam
+    #   nltk
+    #   sacrebleu
 requests==2.32.3
     # via
     #   apache-beam
     #   gcsfs
     #   google-api-core
     #   google-cloud-bigquery
+    #   google-cloud-storage
     #   hdfs
     #   jupyterlab-server
+    #   keyrings-google-artifactregistry-auth
     #   requests-oauthlib
-    #   tensorboard
+    #   tensorflow
 requests-oauthlib==2.0.0
     # via google-auth-oauthlib
 rfc3339-validator==0.1.4
@@ -540,7 +674,11 @@ rfc3986-validator==0.1.1
     # via
     #   jsonschema
     #   jupyter-events
-rpds-py==0.20.1
+rich==13.9.4
+    # via keras
+rouge-score==0.1.2
+    # via tensorflow-model-analysis
+rpds-py==0.23.1
     # via
     #   jsonschema
     #   referencing
@@ -548,86 +686,91 @@ rsa==4.9
     # via
     #   google-auth
     #   oauth2client
-scikit-learn==0.24.2
-    # via -r -
-scipy==1.10.1
+sacrebleu==2.5.1
+    # via tensorflow-model-analysis
+scikit-learn==1.5.0
+    # via -r backend/src/apiserver/visualization/requirements.in
+scipy==1.15.2
     # via
     #   scikit-learn
     #   tensorflow-model-analysis
+secretstorage==3.3.3
+    # via keyring
 send2trash==1.8.3
     # via jupyter-server
+shapely==2.0.7
+    # via google-cloud-aiplatform
 six==1.17.0
     # via
     #   astunparse
-    #   bleach
-    #   bokeh
     #   google-api-python-client
     #   google-apitools
     #   google-pasta
     #   hdfs
-    #   keras-preprocessing
     #   oauth2client
     #   python-dateutil
     #   rfc3339-validator
+    #   rouge-score
+    #   tensorboard
     #   tensorflow
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
 sniffio==1.3.1
     # via anyio
+sortedcontainers==2.4.0
+    # via apache-beam
 soupsieve==2.6
     # via beautifulsoup4
 sqlparse==0.5.3
     # via google-cloud-spanner
-tensorboard==2.10.1
+tabulate==0.9.0
+    # via sacrebleu
+tensorboard==2.16.2
     # via tensorflow
-tensorboard-data-server==0.6.1
+tensorboard-data-server==0.7.2
     # via tensorboard
-tensorboard-plugin-wit==1.8.1
-    # via tensorboard
-tensorflow==2.10.1
+tensorflow==2.16.2
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
     #   tensorflow-serving-api
     #   tfx-bsl
-tensorflow-data-validation==1.9.0
-    # via -r -
-tensorflow-estimator==2.10.0
-    # via tensorflow
-tensorflow-io-gcs-filesystem==0.34.0
+tensorflow-data-validation==1.16.1
+    # via -r backend/src/apiserver/visualization/requirements.in
+tensorflow-estimator==2.15.0
+    # via tensorflow-model-analysis
+tensorflow-io-gcs-filesystem==0.37.1
     # via tensorflow
-tensorflow-metadata==1.9.0
+tensorflow-metadata==1.16.1
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
     #   tfx-bsl
-tensorflow-model-analysis==0.40.0
-    # via -r -
-tensorflow-serving-api==2.10.1
+tensorflow-model-analysis==0.47.1
+    # via -r backend/src/apiserver/visualization/requirements.in
+tensorflow-serving-api==2.14.1
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   tfx-bsl
-termcolor==2.4.0
+termcolor==2.5.0
     # via tensorflow
 terminado==0.18.1
     # via
     #   jupyter-server
     #   jupyter-server-terminals
-tfx-bsl==1.9.0
+tfx-bsl==1.16.1
     # via
     #   tensorflow-data-validation
     #   tensorflow-model-analysis
 threadpoolctl==3.5.0
     # via scikit-learn
-tinycss2==1.2.1
+tinycss2==1.4.0
     # via bleach
-tomli==2.2.1
-    # via jupyterlab
 tornado==6.4.2
     # via
-    #   -r -
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   bokeh
     #   ipykernel
     #   jupyter-client
@@ -635,6 +778,8 @@ tornado==6.4.2
     #   jupyterlab
     #   notebook
     #   terminado
+tqdm==4.67.1
+    # via nltk
 traitlets==5.14.3
     # via
     #   comm
@@ -646,6 +791,7 @@ traitlets==5.14.3
     #   jupyter-events
     #   jupyter-server
     #   jupyterlab
+    #   matplotlib-inline
     #   nbclient
     #   nbconvert
     #   nbformat
@@ -655,19 +801,24 @@ typing-extensions==4.12.2
     # via
     #   anyio
     #   apache-beam
-    #   async-lru
-    #   mistune
-    #   python-json-logger
+    #   beautifulsoup4
+    #   bokeh
+    #   google-cloud-aiplatform
+    #   opentelemetry-sdk
+    #   optree
+    #   pydantic
+    #   pydantic-core
+    #   referencing
     #   tensorflow
 uri-template==1.3.0
     # via jsonschema
 uritemplate==3.0.1
     # via google-api-python-client
-urllib3==2.2.3
+urllib3==2.3.0
     # via requests
 wcwidth==0.2.13
     # via prompt-toolkit
-webcolors==24.8.0
+webcolors==24.11.1
     # via jsonschema
 webencodings==0.5.1
     # via
@@ -675,20 +826,20 @@ webencodings==0.5.1
     #   tinycss2
 websocket-client==1.8.0
     # via jupyter-server
-werkzeug==2.1.2
-    # via tensorboard
-wheel==0.45.1
+werkzeug==3.0.6
     # via
-    #   astunparse
+    #   -r backend/src/apiserver/visualization/requirements.in
     #   tensorboard
+wheel==0.45.1
+    # via astunparse
 widgetsnbextension==3.6.10
     # via ipywidgets
 wrapt==1.17.2
-    # via tensorflow
-zipp==3.20.2
     # via
-    #   importlib-metadata
-    #   importlib-resources
+    #   deprecated
+    #   tensorflow
+zipp==3.21.0
+    # via importlib-metadata
 zstandard==0.23.0
     # via apache-beam
